File: src/auth/parser/servicetemplate.go
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package parser

import (
	"errors"
	"net/http"

	"configcenter/src/auth/meta"
	"configcenter/src/common"

	"github.com/tidwall/gjson"
)

var ServiceTemplateAuthConfigs = []AuthConfig{
	{
		Name:                  "createServiceTemplatePattern",
		Description:           "创建服务模板",
		Pattern:               "/process/v3/create/proc/service_template",
		HTTPMethod:            http.MethodPost,
		RequiredBizInMetadata: true,
		ResourceType:          meta.ProcessServiceTemplate,
		ResourceAction:        meta.Create,
	}, {
		Name:                  "listServiceTemplatePattern",
		Description:           "查询服务模板",
		Pattern:               "/process/v3/findmany/proc/service_template",
		HTTPMethod:            http.MethodPost,
		RequiredBizInMetadata: true,
		ResourceType:          meta.ProcessServiceTemplate,
		ResourceAction:        meta.FindMany,
	}, {
		Name:                  "listServiceTemplateDetailPattern",
		Description:           "查询服务模板详情",
		Pattern:               "/process/v3/findmany/proc/service_template/with_detail",
		HTTPMethod:            http.MethodPost,
		RequiredBizInMetadata: true,
		ResourceType:          meta.ProcessServiceTemplate,
		ResourceAction:        meta.FindMany,
	}, {
		Name:                  "deleteServiceTemplatePattern",
		Description:           "删除服务模板",
		Pattern:               "/process/v3/delete/proc/service_template",
		HTTPMethod:            http.MethodDelete,
		RequiredBizInMetadata: true,
		ResourceType:          meta.ProcessServiceTemplate,
		ResourceAction:        meta.Delete,
		InstanceIDGetter: func(request *RequestContext) (int64s []int64, e error) {
			templateID := gjson.GetBytes(request.Body, common.BKServiceTemplateIDField).Int()
			if templateID <= 0 {
				return nil, errors.New("invalid service template")
			}
			return []int64{templateID}, nil
		},
	},
}

func (ps *parseStream) ServiceTemplate() *parseStream {
	resources, err := MatchAndGenerateIAMResource(ServiceTemplateAuthConfigs, ps.RequestCtx)
	if err != nil {
		ps.err = err
	}
	if resources != nil {
		ps.Attribute.Resources = resources
	}
	return ps
}

File: src/ui/src/store/modules/api/service-template.js
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

import $http from '@/api'

const state = {}

const getters = {}

const actions = {
    /**
     * 查询服务模板详情列表
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @param {Object} params 参数
     * @return {promises} promises 对象
     */
    searchServiceTemplate ({ commit, state, dispatch, rootGetters }, { params, config }) {
        return $http.post(`findmany/proc/service_template/with_detail`, params, config)
    },
    /**
     * 查询服务模板详情列表
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @param {Object} params 参数
     * @return {promises} promises 对象
     */
    searchServiceTemplateWithoutDetails ({ commit, state, dispatch, rootGetters }, { params, config }) {
        return $http.post(`findmany/proc/service_template`, params, config)
    },
    /**
     * 创建服务模板
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @param {Object} params 参数
     * @return {promises} promises 对象
     */
    createServiceTemplate ({ commit, state, dispatch, rootGetters }, { params, config }) {
        return $http.post(`create/proc/service_template`, params, config)
    },
    /**
     * 编辑服务模板
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @param {Object} params 参数
     * @return {promises} promises 对象
     */
    updateServiceTemplate ({ commit, state, dispatch, rootGetters }, { params, config }) {
        return $http.put(`update/proc/service_template`, params, config)
    },
    /**
     * 删除服务模板
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @param {Object} params 参数
     * @return {promises} promises 对象
     */
    deleteServiceTemplate ({ commit, state, dispatch, rootGetters }, { params, config }) {
        return $http.delete(`delete/proc/service_template`, params)
    },
    /**
     * 查看单个服务模板
     * @param {Function} commit store commit mutation hander
     * @param {Object} state store state
     * @param {String} dispatch store dispatch action hander
     * @param {Object} params 参数
     * @return {promises} promises 对象
     */
    findServiceTemplate ({ commit, state, dispatch, rootGetters }, { id, config }) {
        return $http.get(`find/proc/service_template/${id}/detail`, config)
    },

    getServiceTemplateModules (context, { bizId, serviceTemplateId, params, config }) {
        return $http.post(`module/bk_biz_id/${bizId}/service_template_id/${serviceTemplateId}`, params, config)
    }
}

const mutations = {}

export default {
    namespaced: true,
    state,
    getters,
    actions,
    mutations
}
